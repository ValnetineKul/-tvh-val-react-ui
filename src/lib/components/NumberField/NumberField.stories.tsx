import React, { useState } from 'react';
import { action } from '@storybook/addon-actions';
import type { StoryTemplate } from '../../story-utils';
import { createMeta } from '../../story-utils';
import type { NumberFieldProps } from './NumberField';
import NumberField from './NumberField';

const Template = ((args: NumberFieldProps) => {
  const [number, setNumber] = useState(null);

  const handleChange = (value: number) => {
    action('onChange')(value);
    setNumber(value);
  };

  return <NumberField {...args} value={number} onChange={handleChange} />;
}) as StoryTemplate<NumberFieldProps>;

export const Primitive = Template.bind({});
Primitive.args = {
  label: 'Quantity',
};
Primitive.storyName = 'NumberField';

export default createMeta({
  component: NumberField,
  title: 'Components/InputFields/NumberField',
  argTypes: {
    onChange: {
      action: { argTypesRegex: '^on.*' },
    },
    inputRef: {
      table: {
        disable: true,
      },
    },
    value: {
      control: {
        disable: true,
      },
    },
    decimalScale: {
      table: {
        type: {
          summary: 'It defines number of digits that could be added after decimalSeparator',
        },
      },
    },
  },
});
