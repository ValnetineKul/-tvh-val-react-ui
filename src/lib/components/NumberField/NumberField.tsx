import type { InputBaseComponentProps } from '@mui/material';
import type { FC } from 'react';
import React, { useMemo, useCallback, forwardRef } from 'react';

import type { NumberFormatValues } from 'react-number-format';
import TextField from '../TextField';
import type { DataAttributes } from '../../types/common';
import NumberFormatInput from '../../../forms/NumberFormatInput';
import { useConfigContext } from '../../../themes/core';
import getSeparatorsForLocale from './helpers/getSeparatorsForLocale';

export interface NumberFieldProps {
  decimalScale?: number;
  fixedDecimalScale?: boolean;
  id?: string;
  label?: string;
  value?: number | string;
  prefix?: string;
  suffix?: string;
  errorMessage?: string;
  negativeNumberErrorMessage?: string;
  placeholder?: string;
  inputClassName?: string;
  className?: string;
  disabled?: boolean;
  readOnly?: boolean;
  fullWidth?: boolean;
  required?: boolean;
  autoFocus?: boolean;
  autoSelect?: boolean;
  onChange?: (value: number | undefined) => void;
  inputRef?: React.RefObject<HTMLInputElement>;
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  onKeyDown?: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  maxNumberOfDigits?: number;
  inputProps?: InputBaseComponentProps & DataAttributes;
}

const InputComponent = forwardRef((props, ref) => <NumberFormatInput {...props} inputRef={ref as never} />);

const NumberField: FC<NumberFieldProps> = ({
  value,
  errorMessage,
  negativeNumberErrorMessage = 'Negative numbers cannot be used',
  className,
  inputClassName,
  decimalScale,
  fixedDecimalScale = true,
  onChange,
  onBlur,
  onKeyDown,
  maxNumberOfDigits = 15,
  inputProps: incomingInputProps,
  inputRef,
  ...props
}) => {
  const [isNegativeNumber, setIsNegativeNumberError] = React.useState(false);

  const actualDecimalScale = React.useMemo(() => {
    if (decimalScale > 0) return decimalScale;
    return undefined;
  }, [decimalScale]);

  const errorMessageComposed = errorMessage || (isNegativeNumber && negativeNumberErrorMessage);

  const { locale } = useConfigContext();

  const { thousandSeparator, decimalSeparator } = useMemo(() => getSeparatorsForLocale(locale), [locale]);

  const checkRequiredLength = useCallback(
    (numberValue: string) => numberValue.length <= maxNumberOfDigits,
    [maxNumberOfDigits]
  );

  const handleChange = (numberValue: number | undefined) => {
    if (onChange) {
      isNegativeNumber && setIsNegativeNumberError(false);
      onChange(numberValue);
    }
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    event.stopPropagation();
    const { target, key } = event;

    if (key === '-') {
      event.preventDefault();
      return;
    }

    if (key === decimalSeparator) {
      const { value: targetValue, selectionStart } = target as HTMLInputElement;
      const decimalSeparatorPosition = targetValue.length - (actualDecimalScale || 0) - 1;
      if (selectionStart !== decimalSeparatorPosition) {
        event.preventDefault();
      }
      return;
    }

    if (key === 'ArrowUp') {
      if (onChange) {
        let newValue = value || 0;
        if (typeof newValue === 'string') {
          newValue = 0;
        } else {
          newValue += 1;
        }
        onChange(newValue);
      }
      return;
    }
    if (key === 'ArrowDown') {
      if (onChange) {
        let newValue = value || 0;
        if (typeof newValue === 'string' || newValue <= 0) {
          newValue = 0;
        } else {
          newValue -= 1;
        }
        onChange(newValue);
      }
      return;
    }

    if (onKeyDown) {
      onKeyDown(event);
    }
  };

  const handlePasteNegativeDigit = (event: React.ClipboardEvent<HTMLInputElement>) => {
    const { clipboardData } = event;
    const pastedData = clipboardData.getData('text');
    const pastedNumber = parseFloat(pastedData);
    if (pastedData.includes('-')) {
      event.preventDefault();
    }

    if (pastedNumber < 0) {
      event.preventDefault();
      setIsNegativeNumberError(true);
    }

    return null;
  };

  const inputProps: InputBaseComponentProps = {
    decimalScale: actualDecimalScale,
    thousandSeparator,
    decimalSeparator,
    fixedDecimalScale,
    onValueChange: (values: { floatValue: number }) => {
      const { floatValue } = values;
      handleChange(floatValue);
    },
    checkValueIsValid: (numberValue: NumberFormatValues) => checkRequiredLength(numberValue.value),
    ...(incomingInputProps || {}),
  };

  return (
    <TextField
      {...props}
      value={value}
      inputComponent={InputComponent}
      inputProps={inputProps}
      inputClassName={inputClassName}
      className={className}
      errorMessage={errorMessageComposed}
      inputElementRef={inputRef}
      onBlur={onBlur}
      onKeyDown={handleKeyDown}
      onPaste={handlePasteNegativeDigit}
      autoComplete="off"
    />
  );
};

export default NumberField;
