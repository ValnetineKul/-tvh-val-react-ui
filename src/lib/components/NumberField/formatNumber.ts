function applyThousandSeparator(value: string, thousandSeparator: string): string {
  return value.replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator);
}

const formatNumber = (
  value: number,
  decimalScale?: number,
  decimalSeparator?: string,
  thousandSeparator?: string
): string => {
  let formattedValue = String(value);
  if (decimalScale) {
    formattedValue = value.toFixed(decimalScale);
  }
  if (decimalSeparator && decimalSeparator !== '.') {
    formattedValue = formattedValue.replace('.', decimalSeparator);
  }
  if (thousandSeparator) {
    if (decimalScale) {
      const [int, cents] = value.toFixed(decimalScale).split('.');
      formattedValue = applyThousandSeparator(int, thousandSeparator) + decimalSeparator + cents;
    } else {
      formattedValue = applyThousandSeparator(formattedValue, thousandSeparator);
    }
  }
  return formattedValue;
};

export default formatNumber;
