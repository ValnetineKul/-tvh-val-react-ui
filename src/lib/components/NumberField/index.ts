import NumberField from './NumberField';

export type { NumberFieldProps } from './NumberField';
export default NumberField;
