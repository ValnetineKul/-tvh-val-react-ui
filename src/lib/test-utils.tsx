import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
