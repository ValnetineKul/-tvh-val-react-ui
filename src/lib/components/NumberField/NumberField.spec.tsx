import React from 'react';
import { createEvent, fireEvent } from '@testing-library/react';
import { render, userEvent } from '../../test-utils';
import NumberField from './NumberField';

const pasteNumberFromClipboard = (inputElement: HTMLElement, clipboardNumber: string) => {
  const pasteEvent = createEvent.paste(inputElement, {
    clipboardData: {
      getData: () => clipboardNumber,
    },
  });

  fireEvent(inputElement, pasteEvent);

  return pasteEvent;
};

describe('NumberField', () => {
  it('Should render correctly', () => {
    const { container } = render(<NumberField />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Should be empty by default', () => {
    const { getByLabelText } = render(<NumberField label="Quantity" />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    expect(input).toHaveValue('');
  });

  describe('Should have formatted value', () => {
    const cases = [
      [1, '1'],
      [1000, '1,000'],
      [10000, '10,000'],
      [100000, '100,000'],
      [1000000, '1,000,000'],
      [10000000, '10,000,000'],
    ];
    test.each(cases)(
      'given %p, returns %p',
      (value: React.ComponentProps<typeof NumberField>['value'], expected: string) => {
        const { getByLabelText } = render(<NumberField label="Quantity" value={value} />);
        const input = getByLabelText('Quantity', { selector: 'input', exact: false });
        expect(input).toHaveValue(expected);
      }
    );
  });
  describe('Should have formatted float value', () => {
    const cases = [
      [1, '1.00'],
      [1000, '1,000.00'],
      [10000, '10,000.00'],
      [100000, '100,000.00'],
      [1000000, '1,000,000.00'],
      [10000000, '10,000,000.00'],
    ];
    test.each(cases)(
      'given %p, returns %p',
      (value: React.ComponentProps<typeof NumberField>['value'], expected: string) => {
        const { getByLabelText } = render(<NumberField label="Quantity" value={value} decimalScale={2} />);
        const input = getByLabelText('Quantity', { selector: 'input', exact: false });
        expect(input).toHaveValue(expected);
      }
    );
  });
  it('Should trigger onChange', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(<NumberField label="Quantity" onChange={handleChange} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    userEvent.type(input, '42');
    expect(handleChange).toHaveBeenCalledTimes(2);
    expect(input).toHaveValue('42');
  });
  it('Should show error message', () => {
    const { getByText } = render(<NumberField errorMessage="Error message example" />);
    const errorMessage = getByText(/Error message example/);
    expect(errorMessage).toBeInTheDocument();
  });
  it('Should render value with decimals', () => {
    const { getByLabelText, rerender } = render(<NumberField label="Quantity" value={1000} decimalScale={2} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    expect(input).toHaveValue('1,000.00');

    rerender(<NumberField label="Rerender" value={1000} decimalScale={5} />);
    const rerenderInput = getByLabelText('Rerender', { selector: 'input', exact: false });
    expect(rerenderInput).toHaveValue('1,000.00000');
  });
  it('Should not show decimalSeparator', () => {
    const { getByLabelText } = render(<NumberField label="Quantity" value={1000} decimalScale={0} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    expect(input).toHaveValue('1,000');
  });
  it('Should change cursor position if decimalSeparator was added before existing separator', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(
      <NumberField label="Quantity" value={1000} decimalScale={2} onChange={handleChange} />
    );
    const input = getByLabelText('Quantity', { selector: 'input', exact: false }) as HTMLInputElement;
    input.setSelectionRange(5, 5);
    userEvent.type(input, '.');
    expect(input.selectionStart).toBe(6);
    expect(handleChange).toHaveBeenCalledTimes(0);
  });
  it('Should not change cursor position if decimalSeparator was added not before existing separator', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(
      <NumberField label="Quantity" value={1000} decimalScale={2} onChange={handleChange} />
    );
    const input = getByLabelText('Quantity', { selector: 'input', exact: false }) as HTMLInputElement;
    input.setSelectionRange(4, 4);
    userEvent.type(input, '.');
    expect(input.selectionStart).toBe(4);
    expect(handleChange).toHaveBeenCalledTimes(0);
  });
  it('Should trigger change on ArrowUp event', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(<NumberField label="Quantity" value={1} onChange={handleChange} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    userEvent.type(input, '{arrowup}');
    expect(handleChange).toHaveBeenCalledWith(2);
    expect(handleChange).toHaveBeenCalledTimes(1);
  });
  it('Should trigger change on ArrowDown event', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(<NumberField label="Quantity" value={1} onChange={handleChange} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    userEvent.type(input, '{arrowdown}');
    expect(handleChange).toHaveBeenCalledWith(0);
    expect(handleChange).toHaveBeenCalledTimes(1);
  });
  it('Should do nothing on minus event', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(<NumberField label="Quantity" value={1} onChange={handleChange} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    userEvent.type(input, '{minus}');
    expect(handleChange).toHaveBeenCalledTimes(0);
  });
  it('Should not allow non-numeric symbols', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(<NumberField label="Quantity" value={1} onChange={handleChange} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    userEvent.type(input, 'a');
    userEvent.type(input, '-');
    userEvent.type(input, '.');
    expect(handleChange).not.toHaveBeenCalled();
  });
  it('Should not exceed 15 digits', () => {
    const handleChange = jest.fn();
    const { getByLabelText } = render(<NumberField label="Quantity" value={123456789012345} onChange={handleChange} />);
    const input = getByLabelText('Quantity', { selector: 'input', exact: false });
    userEvent.type(input, '6');
    expect(handleChange).toHaveBeenCalledTimes(0);
    expect(input).toHaveValue('123,456,789,012,345');
  });

  describe('Should not allow to use negative digits', () => {
    it('Should show error message when negative digits are pasted from clipboard', () => {
      const { getByLabelText, getByText } = render(<NumberField label="Quantity" />);
      const input = getByLabelText('Quantity', { selector: 'input', exact: false });

      pasteNumberFromClipboard(input, '-5');
      expect(getByText('Negative numbers cannot be used')).toBeInTheDocument();
    });

    it('Should not change the existing value to negative when minus symbol is pasted', () => {
      const { getByLabelText } = render(<NumberField label="Quantity" value={3} />);
      const input = getByLabelText('Quantity', { selector: 'input', exact: false });

      const clipBoardEvent = pasteNumberFromClipboard(input, '-');
      expect(clipBoardEvent.defaultPrevented).toBeTruthy();
      expect(input).toHaveValue('3');
    });

    it('Should not change the existing value to negative when a string containing minus symbol is pasted', () => {
      const { getByLabelText } = render(<NumberField label="Quantity" value={3} />);
      const input = getByLabelText('Quantity', { selector: 'input', exact: false });

      const clipBoardEvent = pasteNumberFromClipboard(input, '+-+');
      expect(clipBoardEvent.defaultPrevented).toBeTruthy();
      expect(input).toHaveValue('3');
    });

    it('Should not allow to set negative digist when using arrowdown keyboard navigation', () => {
      const handleChange = jest.fn();
      const { getByLabelText } = render(<NumberField label="Quantity" value={0} onChange={handleChange} />);
      const input = getByLabelText('Quantity', { selector: 'input', exact: false });

      userEvent.type(input, '{arrowdown}');
      expect(handleChange).toHaveBeenCalledWith(0);
    });

    it('Should disable browser autocomplete for correct arrowUp and arrawDown keyboard navigation', () => {
      const { getByLabelText } = render(<NumberField label="Quantity" />);
      const input = getByLabelText('Quantity', { selector: 'input', exact: false });
      expect(input).toHaveAttribute('autocomplete', 'off');
    });
  });
});
