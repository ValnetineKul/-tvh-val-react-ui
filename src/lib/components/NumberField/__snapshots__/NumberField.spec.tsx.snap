// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberField Should render correctly 1`] = `
<div
  class="MuiFormControl-root tss-16kerj4-TextField-root mui-1nrlq1o-MuiFormControl-root"
>
  <div
    class="tss-bdw984-Surface-root"
  >
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl tss-10yfkph-TextField-inputRoot tss-10-TextField-inputRoot-ref mui-183rr2a-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        autocomplete="off"
        class="tss-epi3qm-TextField-input MuiOutlinedInput-input MuiInputBase-input mui-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        inputmode="numeric"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="tss-tyiogj-TextField-notchedOutline tss-10-TextField-notchedOutline-ref MuiOutlinedInput-notchedOutline mui-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="mui-nnbavb"
        >
          <span
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="tss-19z6szl-TextField-helpers"
  />
</div>
`;
